### PR TITLE
[Ftr]:load the API router config into the local memory

### LINKED ISSUE #19: Requests processor for http-to-http and http-to-dubbo
**What would you like to be added**:
As a developer, I want to create requests processors to process and map the http-to-http and http-to-dubbo requests. 
**Why is this needed**:

### PR DESCRIPTION
**Pull request by williamfeng323**
**What this PR does**:

0. Load the API config into the local memory API discovery service

as described in [this flow chart](https://cdn.nlark.com/yuque/0/2020/png/327432/1601110962319-3bcd8c32-0ae1-4248-8c51-ce7b25c53769.png?x-oss-process=image%2Fresize%2Cw_1492)

1. Implement the API router for the RESTful API server to match the inbound request to proper API in proxy/listener.go.
2. handle invoke the matched API with the inbound mapped parameters.

**Which issue(s) this PR fixes**:

Fixes #19 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/client/dubbo/dubbo_client.go [hunk 1]
```diff
@@ -167,7 +159,7 @@ func (dc *DubboClient) check(key string) bool {
 	}
 }
 
-func (dc *DubboClient) create(key string, dm *DubboMetadata) *dg.GenericService {
+func (dc *DubboClient) create(key string, dm config.IntegrationRequest) *dg.GenericService {
```
Comment (cityiron) [inline pkg/client/dubbo/dubbo_client.go:162]: renam dm to another, dm is mean DubboMetadata.

File: pkg/config/api_config.go [hunk 2]
```diff
@@ -49,81 +54,112 @@ const (
...
+	Schema string `json:"schema" yaml:"schema"` // use json schema
 }
 
 // LoadAPIConfigFromFile load the api config from file
 func LoadAPIConfigFromFile(path string) (*APIConfig, error) {
 	if len(path) == 0 {
 		return nil, perrors.Errorf("Config file not specified")
 	}
-	logger.Info("Load API configuration file form %s", path)
+	logger.Info("Load API configuration file form ", path)
```
[the remark this answers is not in the document]
Comment (williamfeng323) [reply pkg/config/api_config.go:150]: must be a typo. I should not remove tat %s

File: pkg/client/dubbo/dubbo_client.go (hunk 1 above)
Comment (williamfeng323) [reply pkg/client/dubbo/dubbo_client.go:162]: ok.

File: configs/api_config.yaml [hunk 3]
```diff
@@ -8,35 +8,22 @@ resources:
...
           queryStrings:
             - name: id
               required: false
-            - name: type
-              required: false
-          requestBody:
-            - definitionName: modelDefinition
         integrationRequest:
-          requestType: dubbo
+          requestType: DUBBO
```
Comment (zouyx) [inline configs/api_config.yaml:18]: why change it?

File: pkg/client/dubbo/dubbo_client.go [hunk 4]
```diff
@@ -167,29 +159,29 @@ func (dc *DubboClient) check(key string) bool {
...
+	referenceConfig.InterfaceName = irequest.Interface
 	referenceConfig.Cluster = constant.DEFAULT_CLUSTER
 	var registers []string
 	for k := range dgCfg.Registries {
 		registers = append(registers, k)
 	}
 	referenceConfig.Registry = strings.Join(registers, ",")
 
-	if dm.ProtocolTypeStr == "" {
+	if irequest.Protocol == "" {
```
Comment (zouyx) [inline pkg/client/dubbo/dubbo_client.go:172]: len()==0

File: pkg/client/dubbo/dubbo_client.go [hunk 5]
```diff
@@ -167,29 +159,29 @@ func (dc *DubboClient) check(key string) bool {
...
+		referenceConfig.Protocol = irequest.Protocol
 	}
 
-	referenceConfig.Version = dm.Version
-	referenceConfig.Group = dm.Group
+	referenceConfig.Version = irequest.Version
+	referenceConfig.Group = irequest.Group
 	referenceConfig.Generic = true
-	if dm.Retries == "" {
+	if irequest.Retries == "" {
```
Comment (zouyx) [inline pkg/client/dubbo/dubbo_client.go:181]: as above

File: configs/client.yml [hunk 6]
```diff
@@ -3,55 +3,47 @@
...
+  environment : "dev"
 
-registries:
-  "zk_1":
+registries :
+  "hangzhouzk":
     protocol: "zookeeper"
-    timeout: "3s"
-    address: "127.0.0.1:2182"
+    timeout	: "3s"
```
Comment (zhangshen023) [inline configs/client.yml:22]: cut this spaceline

File: pkg/context/http/context.go [hunk 7]
```diff
@@ -36,7 +38,7 @@ type HttpContext struct {
 	HttpConnectionManager model.HttpConnectionManager
 	FilterChains          []model.FilterChain
 	Listener              *model.Listener
-	api                   *model.Api
+	api                   router.API
```
Comment (zhangshen023) [inline pkg/context/http/context.go:41]: why not pointer?

File: pkg/context/http/context_test.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"github.com/dubbogo/dubbo-go-proxy/pkg/common/extension"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/config"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/context"
+	httpCtx "github.com/dubbogo/dubbo-go-proxy/pkg/context/http"
+	_ "github.com/dubbogo/dubbo-go-proxy/pkg/filter"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/model"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/router"
+	"github.com/stretchr/testify/assert"
```
Comment (zhangshen023) [inline pkg/context/http/context_test.go:32]: spilit

File: pkg/proxy/listener.go [hunk 9]
```diff
@@ -28,10 +28,13 @@ import (
 import (
 	"github.com/dubbogo/dubbo-go-proxy/pkg/common/constant"
 	"github.com/dubbogo/dubbo-go-proxy/pkg/common/extension"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/config"
 	"github.com/dubbogo/dubbo-go-proxy/pkg/context"
 	h "github.com/dubbogo/dubbo-go-proxy/pkg/context/http"
 	"github.com/dubbogo/dubbo-go-proxy/pkg/logger"
 	"github.com/dubbogo/dubbo-go-proxy/pkg/model"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/router"
+	"github.com/pkg/errors"
```
Comment (zhangshen023) [inline pkg/proxy/listener.go:37]: spilit

File: pkg/context/http/context.go (hunk 7 above)
Comment (williamfeng323) [reply pkg/context/http/context.go:41]: i don't want the origin api be changed in the context. So I change it to not use the pointer

File: configs/api_config.yaml (hunk 3 above)
Comment (williamfeng323) [reply configs/api_config.yaml:18]: typo here.

File: pkg/proxy/listener.go (hunk 9 above)
Comment (williamfeng323) [reply pkg/proxy/listener.go]: fixed

File: pkg/context/http/context_test.go (hunk 8 above)
Comment (williamfeng323) [reply pkg/context/http/context_test.go]: fixed

File: configs/client.yml (hunk 6 above)
Comment (williamfeng323) [reply configs/client.yml]: fixed

File: pkg/client/dubbo/dubbo_client.go (hunk 5 above)
Comment (williamfeng323) [reply pkg/client/dubbo/dubbo_client.go]: fixed

File: pkg/client/dubbo/dubbo_client.go (hunk 4 above)
Comment (williamfeng323) [reply pkg/client/dubbo/dubbo_client.go]: fixed

File: configs/api_config.yaml (hunk 3 above)
Comment (williamfeng323) [reply configs/api_config.yaml]: fixed

**Review by cityiron — APPROVED** [review]

**Review by zhangshen023 — APPROVED** [review]

File: pkg/proxy/listener_test.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+import (
+	"github.com/dubbogo/dubbo-go-proxy/pkg/common/constant"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/common/extension"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/config"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/context"
+	ctxHttp "github.com/dubbogo/dubbo-go-proxy/pkg/context/http"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/model"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/router"
+	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (williamfeng323) [reply pkg/proxy/listener_test.go:35]: done

File: pkg/proxy/listener_test.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+		OnAir:              true,
+		HTTPVerb:           verb,
+		InboundRequest:     inbound,
+		IntegrationRequest: integration,
+	}
+	return router.API{
+		URLPattern: urlPattern,
+		Method:     method,
+	}
+}
```
[the remark this answers is not in the document]
Comment (williamfeng323) [reply pkg/proxy/listener_test.go:77]: done

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**